plugins/inputs/wireless/wireless_linux.go

```diff
@@ -135,9 +134,3 @@ func proc(env, defaultPath string) string {
 	// return default path
 	return defaultPath
 }
-
-func init() {
-	inputs.Add("wireless", func() telegraf.Input {
-		return &Wireless{}
-	})
-}
```

Comment:
Why do you remove this code? Without it the plugin will not be registered on Linux and thus will not be available... Or am I missing something?

Reply:
See wireless.go

Reply:
It is only required in the `wireless.go` part. The platform-specific ones don't need it.

Reply:
To be honest, I think it should be the other way around. What is the point in registering this plugin for platforms where it is dysfunctional? Say in FreeBSD someone configures this plugin ... Then he or she will also not get any metrics not any hint on what's going on...

Reply:
Yeah maybe, it also helps to put different default settings for different platforms.

But your scenario will end up here: https://github.com/influxdata/telegraf/blob/5adecc3cd9b93252388142b8e155b37d393de8df/plugins/inputs/wireless/wireless_notlinux.go#L12

Reply:
I will wait to resolve this until you folks decide what the resolution should actually be. 

This was, in fact, the original reason for having a separate plugin for macOS. (actually, the _original_ original was a single plugin, but then someone took the linux code out and submitted it as a separate PR, which got merged, so I had to go the macOS-specific route. )

Reply:
I agree with @srebhan that it should be existing in the platform specific `wireless_*.go` files, but not in `wireless.go`. So I don't see what you are waiting for. What more info do you need from us?

Reply:
I am not 100% certain that this will work, but I will test it. 

I was waiting to resolve it because you said:

> Yeah maybe, it also helps to put different default settings for different platforms.

Which wasn't, to me, an agreement. If you're saying that this is the direction you both want to go, then I will build and test it. 

In this case, I do not think that `wireless.go` is needed _at all_ as everything can just be included in the platform-specific files with no need for a generic `wirelesss.go` at all.

Reply:
It's still needed for the generic `Description` and `SampleConfig` methods, maybe also for the struct (not sure of that)

Reply:
Since the `sample config` is different for the 2 platforms, is it needed in a higher level file? Is having another file just for those 2 small methods not a waste of space? The `struct` as well is platform-specific, so should it not also be included in the `wireless_*.go` files rather than the higher-level file? (they both share the `log` field, but that's it.)

I can confirm that moving those few small parts to the platform-specific files, and removing `wireless.go` entirely still works as it should.

I'm really just trying to get this all nailed down once and for all. Initially having a separate plugin for macOS was not ideal, so I integrated it into the existing (linux-only) plugin. Which triggered a decision that the `init()` function wasn't needed in the platform-specific files, and now we're deciding that it is needed in _those_ files but not the generic higher-level file. 

I'm simply trying to minimize the amount of back-and-forth and re-engineering on my part. There has been a *lot* of "change this" ... "ok, never mind, but change this instead" ... "ok, wait, let's do this instead" already. I make changes to methods, etc as asked, only to have the entire issue re-surface again because someone has a different idea. To recap:

1. The original plugin had a `wireless.go`, a `wireless_linux.go` and a `wireless_notlinux.go` file, each of which had their own `init()` method. 
2. In integrating the `mac_wireless` plugin to this one (as requested) I followed that structure exactly (since it was already 'approved').
3. I was asked Why I did that. Which I explained as above. 
4. I was asked to 'try' removing it from the platform-specific files, which I did, and tested it, and found that it worked, so I submitted that.
5. I'm now asked to do it the opposite way and delete the 'init()` method from the higher-level file and re-add it to the platform-specific files. 

So what I'm asking for here is a definitive design decision. Not a "maybe", etc. but an agreement that we should do it _this_ way or _that_ way that we agree on and is clear so that I don't have to keep re-doing this PR. I really don't think I'm asking for a lot in that request. 

I appreciate the time spent on code review. I do. But I don't like the re-engineering on the fly and the "maybe change this and see if it works?" stuff, only to be asked later to "try it this way instead".

Reply:
> To be honest, I think it should be the other way around. What is the point in registering this plugin for platforms where it is dysfunctional? Say in FreeBSD someone configures this plugin ... Then he or she will not get any metrics not any hint on what's going on...

We could remove the `nonsupport` file altogether. Since the `wireless` plugin will *only* compile on either darwin or linux, if someone adds it on BSD there will be no wireless plugin and telegraf will die because of an invalid config file.

Reply:
This is exhausting.

Reply:
David, I'm sorry, but I really don't understand what more do you want from me?

Reply:
@davidgs I do understand your viewpoint, but please also try to see our point: You contribute this plugin which is very much appreciated, but we need to maintain it over a looong time and also deal with all user feedback. In this special case, people might put `wireless` into their config on say FreeBSD or Windows (for whatever reason) and then complain because they neither get an error nor any data. We then need to spend time to figure out what's going on. Would you agree that any time spent **now** is better spent than hunting issues that could have been prevented?

Reply:
> I agree with @srebhan the init method is only needed in the platform specific when it has different default values..

I'm not sure what this means.  "when it has different default values" meaning what? 

```go
type Wireless struct {
	HostProc string          `toml:"host_proc"`
	Log      telegraf.Logger `toml:"-"`
}

var sampleConfig = `
  ## Sets 'proc' directory path
  ## If not specified, then default is /proc
  ## Ignored on macOS/darwin
  # host_proc = "/proc"
`
```

Are the default values for Linux and macOS. The config values are ignored on macOS because they are not needed. 

I have tested the following scenarios:
- init() in all files :white_check_mark:
- init() in top-level only :white_check_mark:
- init() in platform files only :white_check_mark:
- No top-level file at all, just platform files with init() :white_check_mark: (but `make fmt` has error.)

On any platform that is *not* Linux or macOS, the `wireless_nosupport.go` file will be used, which will return the error: 
`w.Log.Warn("Current platform is not supported")`

That error will be returned no matter which files have the `init()` method.  So they will always get an error if they are not on Linux or macOS. 

All of that being said, it sounds like the decision is to have the `init()` method in the platform-specific files and _not_ the main file, so I will make that change.

Reply:
So I made the change and pushed, but the test is still failing on the listing error which makes no sense.